R/utils-ui.R

```diff
@@ -1,12 +1,264 @@
+      cli::make_ansi_style("grey"),
+      cli::style_italic
+    )
+    ui_bullets(c(" " = style_subtle("[Copied to clipboard]")))
+  }
+
+  invisible(x)
+}
+
+# inspired by gargle::gargle_map_cli() and gargle::bulletize()
+usethis_map_cli <- function(x, ...) UseMethod("usethis_map_cli")
+
+#' @export
+usethis_map_cli.default <- function(x, ...) {
+  ui_abort(c(
+    "x" = "Don't know how to {.fun usethis_map_cli} an object of class
+           {.obj_type_friendly {x}}."
+  ))
+}
+
+#' @export
+usethis_map_cli.NULL <- function(x, ...) NULL
+
+#' @export
+usethis_map_cli.character <- function(x,
+                                      template = "{.val <<x>>}",
+                                      .open = "<<", .close = ">>",
+                                      ...) {
+  as.character(glue(template, .open = .open, .close = .close))
+}
+
+ui_pre_glue <- function(..., .envir = parent.frame()) {
+  glue(..., .open = "<<", .close = ">>", .envir = .envir)
+}
+
+bulletize <- function(x, bullet = "*", n_show = 5, n_fudge = 2) {
+  n <- length(x)
+  n_show_actual <- compute_n_show(n, n_show, n_fudge)
+  out <- utils::head(x, n_show_actual)
+  n_not_shown <- n - n_show_actual
+
+  out <- set_names(out, rep_along(out, bullet))
+
+  if (n_not_shown == 0) {
+    out
+  } else {
+    c(out, " " = glue("{cli::symbol$ellipsis} and {n_not_shown} more"))
+  }
+}
+
+# I don't want to do "... and x more" if x is silly, i.e. 1 or 2
+compute_n_show <- function(n, n_show_nominal = 5, n_fudge = 2) {
+  if (n > n_show_nominal && n - n_show_nominal > n_fudge) {
+    n_show_nominal
+  } else {
+    n
+  }
+}
+
+kv_line <- function(key, value, .envir = parent.frame()) {
+  cli::cli_div(theme = usethis_theme())
+
+  key_fmt <- cli::format_inline(key, .envir = .envir)
+
+  # this must happen first, before `value` has been forced
+  value_fmt <- cli::format_inline("{.val {value}}")
+  # but we might actually want something other than value_fmt
+  if (is.null(value)) {
+    value <- ui_special()
+  }
+  if (inherits(value, "AsIs")) {
+    value_fmt <- cli::format_inline(value, .envir = .envir)
+  }
+
+  ui_bullets(c("*" = "{key_fmt}: {value_fmt}"))
+}
+
+ui_special <- function(x = "unset") {
+  force(x)
+  I(glue("{cli::col_grey('<[x]>')}", .open = "[", .close = "]"))
+}
+
+# errors -----------------------------------------------------------------------
+ui_abort <- function(message, ..., class = NULL, .envir = parent.frame()) {
+  cli::cli_div(theme = usethis_theme())
+
+  nms <- names2(message)
+  default_nms <- rep_along(message, "i")
+  default_nms[1] <- "x"
+  nms <- ifelse(nzchar(nms), nms, default_nms)
+  names(message) <- nms
 
-usethis_abort <- function(message, ..., class = NULL, .envir = parent.frame()) {
-  #cli::cli_div(theme = usethis_theme())
   cli::cli_abort(
     message,
     class = c(class, "usethis_error"),
     .envir = .envir,
     ...
   )
 }
+
+# questions --------------------------------------------------------------------
+ui_yep <- function(x,
```

Comment:
Very similar to `ui_yeah()` but cli is used for preparing `utils:menu(title =, choices =)`.